### PR TITLE
Adding Multinomial Coefficients

### LINKED ISSUE #18933: Multinomial coefficients
I think that multinomial coefficient can be added
It won't be difficult, but the conflicting notations between mathematica and maple is making it confusing.

Some people prefer to use
- image

while some other people prefer to use
- image

assuming that n is loosely the summation of each k.

So it should be chosen that which kind if definition is the best
And if the second one is chosen, I think that I should rather make it like 
image mechanically, so it won't really be much convenient than choosing the first one. But this can be consistent with the definition of binomial.

Some references:
- https://en.wikipedia.org/wiki/Multinomial_theorem
- https://dlmf.nist.gov/26.4
- http://functions.wolfram.com/GammaBetaErf/Multinomial/
- https://www.maplesoft.com/support/help/Maple/view.aspx?path=combinat%2Fmultinomial
- https://reference.wolfram.com/language/ref/Multinomial.html

### PR DESCRIPTION
**Pull request by Arpan612**
Adding Multinomial Coefficients
#### References to other Issues or PRs

Fixes #18933
Closes #18960

#### Brief description of what is fixed or changed
A class for calculating the multinomial coefficient is added here.

#### Release Notes

* functions
   * class for calculating multinomial coefficient is added

### CODE REVIEW

File: sympy/functions/combinatorial/factorials.py
Comment (sylee957) [inline sympy/functions/combinatorial/factorials.py]: You should put this into `eval`

Comment (sylee957) [reply sympy/functions/combinatorial/factorials.py]: And I think that this should be taken as a n-ary function

Comment (Arpan612) [reply sympy/functions/combinatorial/factorials.py]: Okay, sir. Working on it.

**Comment by sylee957** [comment]
And you should only compute the function numerically only when given numeric arguments 
`multinomial(1, 2, 3) -> some value`. 
And other than that, it should rather give a unevaluated form 
`multinomial(a, b, c, d) -> multinomial(a, b, c, d)`
However, you should add some basic algorithms like sorting
`multinomial(c, a, d, b) -> multinomial(a, b, c, d)`

**Comment by Arpan612** [comment]
> And you should only compute the function numerically only when given numeric arguments
> `multinomial(1, 2, 3) -> some value`.
> And other than that, it should rather give a unevaluated form
> `multinomial(a, b, c, d) -> multinomial(a, b, c, d)`
> However, you should add some basic algorithms like sorting
> `multinomial(c, a, d, b) -> multinomial(a, b, c, d)`

Could you tell me how to implement this?

File: sympy/functions/combinatorial/factorials.py
Comment (Arpan612) [reply sympy/functions/combinatorial/factorials.py]: > And I think that this should be a function that should take variable number of arguments

Sir, do you mean make an array of variable arguments?

File: sympy/functions/combinatorial/factorials.py
Comment (czgdp1807) [inline sympy/functions/combinatorial/factorials.py]:
```suggestion
class multinomial(CombinatorialFunction):
```

**Comment by Arpan612** [comment]
@sylee957 Could you please review my PR?

**Comment by Arpan612** [comment]
@czgdp1807 Could you tell me how to squash all commits into one?

**Comment by smitgajjar** [comment]
`git rebase -i HEAD~n`

where, `n` is the number of commits in the branch(or to be squashed)

Replace pick with squash in all the commits except first commit

**Comment by Arpan612** [comment]
> `git rebase -i HEAD~n`
> 
> where, `n` is the number of commits in the branch(or to be squashed)
> 
> Replace pick with squash in all the commits except first commit

The problem I am facing is that some of the commits are pulling changes from master. So I don't know the value of n. Any command which takes all commits into one irrespective of n?

**Comment by Arpan612** [comment]
@czgdp1807 Please review my PR.

File: sympy/functions/combinatorial/factorials.py
Comment (Smit-create) [inline sympy/functions/combinatorial/factorials.py]: Please move the imports at the top of the file.

**Comment by smitgajjar** [comment]
As far as I know, it is advisable not to pull changes from master after submitting a PR. In this case, you can squash it till the commit where you carried out the pull.

**Comment by Arpan612** [comment]
> As far as I know, it is advisable not to pull changes from master after submitting a PR. In this case, you can squash it till the commit where you carried out the pull.

Thank you

File: sympy/functions/combinatorial/factorials.py
Comment (iammosespaulr) [inline sympy/functions/combinatorial/factorials.py]: @Arpan612 dude if you're going to plagiarize my code atleast don't copy it with the errors 😞 ... @sylee957 you might wanna look into this ...

File: sympy/functions/combinatorial/factorials.py
Comment (Smit-create) [inline sympy/functions/combinatorial/factorials.py]: Move these imports at the top of the file and IMO there should be a `return` statement at line 1110

File: sympy/functions/combinatorial/factorials.py
Comment (Smit-create) [inline sympy/functions/combinatorial/factorials.py]: These docstrings are still incorrect. Please have a look and correct it accordingly. You have neither changed the docstrings of `Examples` sections which are still unchanged

**Review by Smit-create — CHANGES_REQUESTED** [review]

File: sympy/functions/combinatorial/factorials.py
Comment (iammosespaulr) [reply sympy/functions/combinatorial/factorials.py]: that's because he plagiarized my code 🤣 ...

**Comment by Arpan612** [comment]
@czgdp1807 Sir, could you please help me out?

**Comment by Smit-create** [comment]
There should be a release note entry for this PR

File: sympy/functions/combinatorial/factorials.py
Comment (czgdp1807) [inline sympy/functions/combinatorial/factorials.py]: Why? This change doesn't appear to be relevant for this PR.

Comment (Arpan612) [reply sympy/functions/combinatorial/factorials.py]: Done sir. Anything else?

**Comment by czgdp1807** [comment]
Tests should be added to enhance the coverage.
P.S.:-
The commits in #18960 for `_eval_*` appeared before here. IMHO, #18960 should be merged. I am attaching `Could Close` label in favour of #18960 
However, if someone(maintainer) thinks that this should go in, feel free to remove the label. 
Thanks.

**Comment by Arpan612** [comment]
@czgdp1807 Sir, I have working in this issue longer than #18960. Please don't add the label.

**Comment by Arpan612** [comment]
I lost the previous issue as well in a similar manner. In this way, he will intentionally pick my issues and work faster to undermine me. Please don't encourage this behaviour.

**Comment by czgdp1807** [comment]
> Please don't add the label.

It's based on my personal opinion. I am not closing this PR because both #18960 and this are really close to each other so timing of commits is the only option left to remove redundancy. Let's see what other maintainers have to say. There is no fixed policy which I am aware of in such cases.

**Comment by Arpan612** [comment]
I have claimed the issue first. I have sent the first PR. Give me just a few hours to correct my code. If I am unable to, then please add the label.

**Comment by czgdp1807** [comment]
Sure, take your time to do things, at the end, IMHO, best PR should be merged. There is nothing to worry about. For me, concrete works matter more. Let's not fill the thread of this PR with unnecessary comments which will not lead to improvement in the code.

**Comment by Arpan612** [comment]
@czgdp1807 Sure sir. I have made all the changes you suggested both here and the PR there. Please review it once.

**Comment by czgdp1807** [comment]
Please add tests for increasing the diff coverage.

**Comment by Arpan612** [comment]
> Please add tests for increasing the diff coverage.

Okay sir. Working on it.

**Comment by czgdp1807** [comment]
The commit history here doesn't look good. Squashing may not work and will be too mechanical perform. You can try the following steps,

0. git checkout -b issue.18933.backup
1. git checkout master
2. git pull <remote pointing to https://github.com/sympy/sympy> master
3. Resolve conflicts if they appear.
4. git checkout issue.18933
5. git merge master
6. Resolve conflicts if they appear.
7. git diff master > ~/sympy_diff.diff
8. git checkout master
9. git branch -D issue.18933
10. git checkout -b issue.18933
11. git apply ~/sympy_diff.diff
12. git push --force <remote pointing to your fork> issue.18933

If you somehow loose your changes due to the above steps, just pick them up from `issue.18933.backup` branch which is suggested in 0th step.

**Comment by Arpan612** [comment]
@czgdp1807 Is this format fine? If yes, I will add 8 more cases like this

**Comment by Arpan612** [comment]
@czgdp1807 Yes sir, got it. As soon as you confirm my code ready to be merged, I will do this.

**Comment by czgdp1807** [comment]
Yes, it looks fine to me.

**Comment by Arpan612** [comment]
@czgdp1807 Great. I will add the cases and squash the commit now.

File: sympy/functions/combinatorial/tests/test_comb_factorials.py
Comment (czgdp1807) [inline sympy/functions/combinatorial/tests/test_comb_factorials.py]: What happens if pass symbols to `multinomial` like, `k1`, `k2`, `k3`?

Comment (czgdp1807) [reply sympy/functions/combinatorial/tests/test_comb_factorials.py]: How `multinomial.rewrite` works?

Comment (Arpan612) [reply sympy/functions/combinatorial/tests/test_comb_factorials.py]: Do you want a test case where we use variables?

Comment (iammosespaulr) [reply sympy/functions/combinatorial/tests/test_comb_factorials.py]: do look at my [tests](https://github.com/sympy/sympy/pull/18960/commits/e953696d24cdeded902fb9e6e4ddba3f7f989b90)

Comment (czgdp1807) [reply sympy/functions/combinatorial/tests/test_comb_factorials.py]: > Do you want a test case where we use variables?

Yes, kind of.

File: sympy/functions/combinatorial/tests/test_comb_factorials.py
Comment (czgdp1807) [inline sympy/functions/combinatorial/tests/test_comb_factorials.py]: It looks like copy pasted from #18960

Comment (czgdp1807) [reply sympy/functions/combinatorial/tests/test_comb_factorials.py]: It would be better you would have cherry-picked the commit if you want to reuse something from other branch/PR.

Comment (Arpan612) [reply sympy/functions/combinatorial/tests/test_comb_factorials.py]: I will make the changes.

Comment (Arpan612) [reply sympy/functions/combinatorial/tests/test_comb_factorials.py]: Could you please have a look at the issues travis ci throws?

**Comment by Arpan612** [comment]
@Smit-create Could you help me out with the bugs?

**Comment by Smit-create** [comment]
Instead of pushing too many commits individually, you should first run the tests locally, check if it passes, then push it upstream. As using Travis is a limited resource provided to the library

**Comment by Arpan612** [comment]
@Smit-create I will squash the commits at the end. Could you help me with the issue travis ci is showing?

**Comment by Arpan612** [comment]
@czgdp1807 @sylee957 Sir, I think all required changes are made. Could you please review it?

**Comment by Arpan612** [comment]
I have also squashed the commits.

File: sympy/functions/combinatorial/factorials.py
Comment (sylee957) [inline sympy/functions/combinatorial/factorials.py]: Same as https://github.com/sympy/sympy/pull/18960/files#r398531441

Comment (Arpan612) [reply sympy/functions/combinatorial/factorials.py]: Done. Could you tell me why is multinomial not defined appearing in travis ci?

**Comment by Arpan612** [comment]
@sylee957 Is the change fine?

File: sympy/__init__.py
Comment (Smit-create) [inline sympy/__init__.py]: Don't make changes that are not related.

File: sympy/functions/combinatorial/factorials.py
Comment (Smit-create) [inline sympy/functions/combinatorial/factorials.py]: This should not be present here. Please refer other docstrings present in the module. Also, these will fail `doctest` as you haven't imported anything in the `Examples `Section

File: sympy/functions/combinatorial/tests/test_comb_factorials.py
Comment (Smit-create) [inline sympy/functions/combinatorial/tests/test_comb_factorials.py]: `k` is defined, but not used. Remove it

File: sympy/functions/combinatorial/tests/test_comb_factorials.py
Comment (Smit-create) [inline sympy/functions/combinatorial/tests/test_comb_factorials.py]: Instead of commenting, remove it

File: sympy/functions/combinatorial/factorials.py
Comment (Smit-create) [inline sympy/functions/combinatorial/factorials.py]: Imports at the top of the file

Comment (Arpan612) [reply sympy/functions/combinatorial/factorials.py]: It is throwing an error.

**Comment by Smit-create** [comment]
Please run the tests locally to avoid too many commits being run on Travis.
Use the following in your branch
```
>>> import sympy
>>> sympy.test('filepath') # this will show you errors locally
>>> sympy.doctest('filepath') # this will show errors in docstrings
```

**Comment by Smit-create** [comment]
Also, add the release note entry.

**Comment by Saanidhyavats** [comment]
@Arpan612 you can go through the following link for running the tests 
https://github.com/sympy/sympy/wiki/Running-tests

File: sympy/functions/combinatorial/factorials.py
Comment (Smit-create) [inline sympy/functions/combinatorial/factorials.py]: You have imported `expand_func`, `Rational`, but haven't used. So, I suggest to remove.

**Comment by Arpan612** [comment]
@moderators Added the small change in doctest suggested by @Smit-create. Tests will pass in about an hour.

**Comment by Arpan612** [comment]
@sylee957 @czgdp1807 @asmeurer Please review and merge this PR.

File: sympy/functions/combinatorial/factorials.py
Comment (Smit-create) [reply sympy/functions/combinatorial/factorials.py]: You haven't made this change.

Comment (Arpan612) [reply sympy/functions/combinatorial/factorials.py]: I have used them. Please check.

Comment (Smit-create) [reply sympy/functions/combinatorial/factorials.py]: Okay.

**Comment by czgdp1807** [comment]
Comment, https://github.com/sympy/sympy/pull/18960#issuecomment-604881417 should be addressed here too.

**Comment by Arpan612** [comment]
@czgdp1807 Can you share the link of the red tests? I am unable to access.

**Comment by czgdp1807** [comment]
https://codecov.io/gh/sympy/sympy/pull/18952/diff

**Comment by Arpan612** [comment]
@czgdp1807 Due to some issue, the page is not loading. I have made as many changes as I could. Can you please check the changes and tell me if anything else is needed?

**Comment by sylee957** [comment]
I was about to check the timestamps of the commits to see who originally had made changes before the other, but I cannot see now because you had squashed all the commits, so I think that you should not edit the commit history before things are decided, or you would lose some evidence.
(If you were sensitive about who’s going to be credited for)

**Comment by czgdp1807** [comment]
I have checked them when the commits in the two PRs were made, #18960 has original commits. So, I will support merging of #18960
Let's wait for the coverage report.

**Comment by Arpan612** [comment]
@sylee957 Sure sir. I will keep that in mind. I was merely following the advice @czgdp1807 sir gave about squashing commits to keep the PR neat.
Also, sir, as stated in the open PR rules, only after 3 days of inactivity, can another person work on an issue. I have not only been active but completed my work in 3 days. I think my PR should be merged.
https://github.com/sympy/sympy/issues/17683

**Comment by Arpan612** [comment]
@czgdp1807 Sir, most of the work has been done by me. I am ready to give credit to #18960 if you say so, but I believe the rules show that my PR could be given first preference.

**Comment by czgdp1807** [comment]
Well, more tests are needed to complete this PR to increase the diff coverage. Please add them.

**Comment by Arpan612** [comment]
@czgdp1807 Sure, sir. Working on it.

File: sympy/functions/combinatorial/factorials.py
Comment (smichr) [inline sympy/functions/combinatorial/factorials.py]: Please don't reproduce wikipedia entries here. Simply state what it does and refer the reader to the [article](https://en.wikipedia.org/wiki/Multinomial_theorem).

File: sympy/functions/combinatorial/factorials.py
Comment (smichr) [inline sympy/functions/combinatorial/factorials.py]: is the condition that `numer > 0` or that all values must be greater than 0?

Comment (sylee957) [reply sympy/functions/combinatorial/factorials.py]: I'd link to this thread https://github.com/sympy/sympy/issues/18933#issuecomment-604405177

File: sympy/functions/combinatorial/factorials.py
Comment (smichr) [inline sympy/functions/combinatorial/factorials.py]: This should allow for fuzzy logic (True, False, None), something like `return fuzzy_group(x.is_integer for x in k)` but check the details of `fuzzy_group`. It could be `fuzzy_and` that is needed.

File: sympy/functions/combinatorial/tests/test_comb_factorials.py
Comment (smichr) [inline sympy/functions/combinatorial/tests/test_comb_factorials.py]: The tests are excessive for the code that was added. Just use symbolic values and check that the symbolic result is correct, then a couple of numeric cases. Any test with symbols that reduces to a number is only testing the already tested gamma functions.

Comment (Arpan612) [reply sympy/functions/combinatorial/tests/test_comb_factorials.py]: Okay sir.

File: sympy/functions/combinatorial/factorials.py
Comment (Arpan612) [reply sympy/functions/combinatorial/factorials.py]: Sir, I think numer > 0 works fine for this case

File: sympy/functions/combinatorial/tests/test_comb_factorials.py
Comment (Arpan612) [reply sympy/functions/combinatorial/tests/test_comb_factorials.py]: Sir, actually @czgdp1807 had asked me to add more tests for increase in code coverage.

File: sympy/functions/combinatorial/factorials.py
Comment (Arpan612) [reply sympy/functions/combinatorial/factorials.py]: I am continuously getting an error when I am trying to implement it. Could you please commit the changes?

**Comment by iammosespaulr** [comment]
> I was about to check the timestamps of the commits to see who originally had made changes before the other, but I cannot see now because you had squashed all the commits, so I think that you should not edit the commit history before things are decided, or you would lose some evidence.
> (If you were sensitive about who’s going to be credited for)

@sylee957 @czgdp1807  you might want to look at my commit history as this person seems to be systematically plagiarizing my code 😅 ... and I believe that it strictly violates this community's ethics.

**Comment by Arpan612** [comment]
No I haven't. We both are doing the same thing of using the existing binomial function and working on it's basis. :+1:

**Comment by smichr** [comment]
I don't like looking at two PRs and duplicating efforts. If there are two different approaches being taken it is worth having two PRs for people to look at, but if they are essentially the same then closure of the one that was opened second (until the first has stalled) would be an appropriate response. It is also possible for collaboration by opening a PR against an existing PR.

File: sympy/functions/combinatorial/tests/test_comb_factorials.py
Comment (czgdp1807) [reply sympy/functions/combinatorial/tests/test_comb_factorials.py]: Tests should have been added in such  a way that uncovered is tested. Adding more tests doesn't necessarily mean that the coverage will increase.

**Comment by czgdp1807** [comment]
Considering all factors, I am closing this PR in favour of #18960 
This has to be done as the efforts are being duplicated and the coverage of both the diffs are same so closing this as per the timestamp of commits. 
Thanks for the contributions and apologies for inconveniences caused if any.

**Comment by smichr** [comment]
> per the timestamp of commits

Is this the right metric? It appears that this PR was opened 2 days ago while #18960 was opened yesterday.

**Comment by smichr** [comment]
(but I am not up to speed with any accusation of plagiarism of one vs the other)

**Comment by czgdp1807** [comment]
> Is this the right metric? It appears that this PR was opened 2 days ago while #18960 was opened yesterday.

When the commits in this PR were not squashed I checked that when #18960 was opened the, commits here started copying code from those commits in #18952 without cherry picking.
Anyways, it was required to close one of the PRs as efforts were duplicated. IMHO, since this conflict is resolved and both the authors will be listed in the `AUTHORS` file, one PR won't make much of a difference to both of them. Investing time and resources on figuring out who is right and who is wrong won't be really helpful.
P.S. To handle such situations in a better way in future, I have proposed two policies in https://github.com/sympy/sympy/issues/17683#issuecomment-604294730
Please let me know if we can make any modifications to them.

**Pull request closed without merging**
